src/combinatorics/simple_graph/trails.lean

```diff
@@ -0,0 +1,168 @@
+      { rw [nat.even_add_one, ih, ← not_iff_not],
+        simp only [huv.ne.symm, ne.def, eq_self_iff_true, not_true, false_and, not_forall,
+          not_false_iff, exists_prop, and_true, not_not, true_and, iff_and_self],
+        rintro rfl,
+        exact huv.ne, } },
+    { rw not_or_distrib at h,
+      simp only [h.1, h.2, not_false_iff, true_and, add_zero, ne.def] at ih ⊢,
+      rw ih,
+      split;
+      { rintro h' h'' rfl,
+        simp only [imp_false, eq_self_iff_true, not_true, not_not] at h',
+        cases h',
+        simpa using h } } },
+end
+
+/-- An *Eulerian trail* (also known as an "Eulerian path") is a walk
+`p` that visits every edge exactly once.  The lemma `simple_graph.walk.is_eulerian.is_trail` shows
+that these are trails.
+
+Combine with `p.is_circuit` to get an Eulerian circuit (also known as an "Eulerian cycle"). -/
+def is_eulerian {u v : V} (p : G.walk u v) : Prop :=
+∀ e, e ∈ G.edge_set → p.edges.count e = 1
+
+lemma is_eulerian.is_trail {u v : V} {p : G.walk u v}
+  (h : p.is_eulerian) : p.is_trail :=
+begin
+  rw [is_trail_def, list.nodup_iff_count_le_one],
+  intro e,
+  by_cases he : e ∈ p.edges,
+  { exact (h e (edges_subset_edge_set _ he)).le },
+  { simp [he] },
+end
+
+lemma is_eulerian.mem_edges_iff {u v : V} {p : G.walk u v} (h : p.is_eulerian) {e : sym2 V} :
+  e ∈ p.edges ↔ e ∈ G.edge_set :=
+⟨λ h, p.edges_subset_edge_set h, λ he, by simpa using (h e he).ge⟩
+
+/-- The edge set of an Eulerian graph is finite. -/
+def is_eulerian.fintype_edge_set {u v : V} {p : G.walk u v}
+  (h : p.is_eulerian) : fintype G.edge_set :=
+fintype.of_finset h.is_trail.edges_finset $ λ e,
+by simp only [finset.mem_mk, multiset.mem_coe, h.mem_edges_iff]
+
+lemma is_trail.is_eulerian_of_forall_mem {u v : V} {p : G.walk u v}
+  (h : p.is_trail) (hc : ∀ e, e ∈ G.edge_set → e ∈ p.edges) :
+  p.is_eulerian :=
+λ e he, list.count_eq_one_of_mem h.edges_nodup (hc e he)
+
+lemma is_eulerian_iff {u v : V} (p : G.walk u v) :
+  p.is_eulerian ↔ p.is_trail ∧ ∀ e, e ∈ G.edge_set → e ∈ p.edges :=
+begin
+  split,
+  { intro h,
+    exact ⟨h.is_trail, λ _, h.mem_edges_iff.mpr⟩, },
+  { rintro ⟨h, hl⟩,
+    exact h.is_eulerian_of_forall_mem hl, },
+end
+
+lemma is_eulerian.edges_finset_eq [fintype G.edge_set]
+  {u v : V} {p : G.walk u v} (h : p.is_eulerian) :
+  h.is_trail.edges_finset = G.edge_finset :=
+by { ext e, simp [h.mem_edges_iff] }
+
+lemma is_eulerian.even_degree_iff {x u v : V} {p : G.walk u v} (ht : p.is_eulerian)
+  [fintype V] [decidable_rel G.adj] :
+  even (G.degree x) ↔ (u ≠ v → x ≠ u ∧ x ≠ v) :=
+begin
+  convert ht.is_trail.even_countp_edges_iff x,
+  rw [← multiset.coe_countp, multiset.countp_eq_card_filter, ← card_incidence_finset_eq_degree],
+  change multiset.card _ = _,
+  congr' 1,
+  convert_to _ = (ht.is_trail.edges_finset.filter (has_mem.mem x)).val,
+  rw [ht.edges_finset_eq, G.incidence_finset_eq_filter x],
+end
+
+lemma is_eulerian.card_filter_odd_degree [fintype V] [decidable_rel G.adj]
```

Comment:
Even though `fin_cases` should unfold this into a theorem of the form `x ∈ 0 ∨ x ∈ 2`, I still believe this should be the preferred way to state this. I recommend that you make
```lean
((finset.univ : finset V).filter (λ v, odd (G.degree v))).card
```
into a separate definition (`odd_degree_count` maybe), prove the equality with `fintype.card {v : V | odd (G.degree v)}`, and then combine this and the below theorem into a single one of the form `odd_degree_count = 0 ∨ odd_degree_count = 2`,

Reply:
I see why you suggest a separate definition, but that can make it harder to apply (or harder to notice it can be applied). I just pushed an experimental design that saves a `convert` in the final lemma -- rather than using a definition, the lemma takes an additional finset and an equality. I expect it to be easier to apply in general. Thanks for pointing this out.

(I find `finset` terms to be very focused on *how* the finite set is constructed. This seems like a good design to me since it makes it easier for the user to specify any finset that has the same elements, decoupling the *how* from the *what*. Not sure if that makes sense.)